rate limit the requests to github

TheAlphaNerd opened the pull request:
github has implemented a new ddos protection method that is causing
consistent failures and timeouts when making more than 50 requests.

This change introduces async and uses its queue functionality and
some setTimeouts with magic numbers to force a rate limit on the
network requests, the result seems to potentially be an overall faster
experience for lower numbers of changes being compared.

Currently async is queuing 15 items and applying a 30ms delay for
callbacks. These numbers could be tunes or potentially exposed as
variables.

rvagg commented on collect-commit-labels.js:
```diff
@@ -19,26 +19,26 @@ function collectCommitLabels (list, callback) {
         if (err) {
           console.error('Error fetching issue #%s: %s', commit.ghIssue, err.message );
-          return done()
+          return setTimeout(next(), 30)
```
`next()` or `next`?

rvagg commented on collect-commit-labels.js:
```diff
@@ -19,26 +19,26 @@ function collectCommitLabels (list, callback) {
         if (issue.labels)
           commit.labels = issue.labels.map(function (label) { return label.name })
-        done()
+        setTimeout(next(), 30)
```
ditto

rvagg commented on collect-commit-labels.js:
```diff
@@ -19,26 +19,26 @@ function collectCommitLabels (list, callback) {
-
-    sublist.forEach(function (commit) {
+    var count = 0
+    var q = async.queue(function(commit, next) {
```
space between `function` and `(`

rvagg commented:
I'm fine with this but do you have a reference for the change the github has made? I can't find anything and it seems a bit strange to not announce a change like this, particularly for authenticated users.

TheAlphaNerd commented:
Updated based on above notes. It was quite late and I'm not sure why I executed the functions when passing to setTimeout. Since things were working without the timeout I've removed it.

AFAIK I have not seen any announcement about this. About 3 days ago I started getting constant time outs when using the tool, perhaps it had to do with the delta of v4 growing so large. I do know that @evanlucas had similar problems and rather than getting a timeout his warning stated something along the lines of his activity being suspicious. 

I had written @jlord about the issues I was having and she was not aware of any specific changes. So obviously this is speculation, but I have never seen these failures before so I'm not really sure what to make of it

TheAlphaNerd commented:
@rvagg thoughts on this? I don't really know how to get any word from github. Either way I don't see this kind of local rate limiting too problematic... I've found the tool to be faster with it 
¯\\\_(ツ)\_/¯.

rvagg commented:
@TheAlphaNerd I still haven't experienced this, but as the heaviest user of changelog-maker and branch-diff at the moment I'll take your word for it that (a) this is a problem and (b) this is the appropriate fix. It works so I've merged and pushed an update with this, v2.2.4.

TheAlphaNerd commented:
thanks @rvagg 🎉
